akd_local_auditor/src/storage/s3.rs

```diff
@@ -0,0 +1,180 @@
+    cache: Arc<RwLock<Option<Vec<super::EpochSummary>>>>,
+}
+
+impl From<&S3ClapSettings> for S3AuditStorage {
+    fn from(clap: &S3ClapSettings) -> Self {
+        Self {
+            bucket: clap.bucket.clone(),
+            config: Arc::new(RwLock::new(None)),
+            cache: Arc::new(RwLock::new(None)),
+        }
+    }
+}
+
+impl S3AuditStorage {
+    async fn get_config(&self) -> s3::Config {
+        let mut lock = self.config.write().await;
+        let shared_config = if let Some(config) = &*lock {
+            config.clone()
+        } else {
+            let sc = aws_config::load_from_env().await;
+            *lock = Some(sc.clone());
+            sc
+        };
+        s3::config::Builder::from(&shared_config)
+            .retry_config(RetryConfig::disabled())
+            .build()
+    }
+
+    async fn list_objects_v2(&self) -> s3::paginator::ListObjectsV2Paginator {
+        let config = self.get_config().await;
+
+        s3::Client::from_conf(config)
+            .list_objects_v2()
+            .bucket(self.bucket.clone())
+            .encoding_type(s3::model::EncodingType::Url)
+            .fetch_owner(false)
+            .into_paginator()
+    }
+
+    async fn get_object(&self, key: &str) -> s3::client::fluent_builders::GetObject {
+        let config = self.get_config().await;
+
+        s3::Client::from_conf(config)
+            .get_object()
+            .bucket(self.bucket.clone())
+            .key(key.clone())
+            .checksum_mode(s3::model::ChecksumMode::Enabled)
+    }
+}
+
+#[async_trait]
+impl super::AuditProofStorage for S3AuditStorage {
+    async fn list_proofs(&self) -> Result<Vec<super::EpochSummary>> {
+        use tokio_stream::StreamExt;
+
+        let mut results = vec![];
+        let client = self.list_objects_v2().await;
+        let mut stream = client.page_size(1000).send();
+        while let Some(result) = stream.next().await {
+            match result {
+                Err(some_error) => {
+                    error!("Error executing list_objects_v2 in S3 {}", some_error);
+                    bail!("Error executing list_objects_v2 in S3 {}", some_error);
+                }
+                Ok(objects) => {
+                    if let Some(contents) = objects.contents() {
+                        for obj in contents {
+                            if let Some(key) = obj.key() {
+                                let summary: Result<super::EpochSummary> = key.try_into();
+                                match summary {
+                                    Ok(dbi) => results.push(dbi),
+                                    Err(error) => {
+                                        debug!("Error parsing blob key into DecomposedBlobItem ({}). Skipping.", error);
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Nit: This is a wee-bit intense. Is there a way we can break this up into something a bit more readable?